### PR TITLE
fix cloud event flaky unit tests by adding waitgroup to fakeclient

### PR DESCRIPTION
**Pull request by Yongxuanzhang**
# Changes

This commit adds waitgroup to fakeclient to avoid that some goroutines are not done when we want to collect the events.
The tests are flaky because the cloud events are sent with goroutine
but we don't wait until all goroutines done to check the events. So it
is possible that some events are not collected. The waitGroup will count when each goroutine is created and decrease the count when the goroutine is done.
This change has no impact on current code.

/kind flake

Signed-off-by: Yongxuanzhang yongxuanzhang@google.com

related issues:
https://github.com/tektoncd/pipeline/issues/5160
to reproduce this issue, adding time.Sleep(100 * time.Millisecond) to https://github.com/tektoncd/pipeline/blob/b648a5b2eaebf0ece7199dcf2bd89c65137973d5/pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go#L53  
related PR:
https://github.com/tektoncd/pipeline/pull/5313
Recent failing tests:

- https://prow.tekton.dev/view/gs/tekton-prow/pr-logs/pull/tektoncd_pipeline/5666/pull-tekton-pipeline-unit-tests/1582795530906374144
- https://prow.tekton.dev/view/gs/tekton-prow/pr-logs/pull/tektoncd_pipeline/5662/pull-tekton-pipeline-unit-tests/1583441668340715520
- https://prow.tekton.dev/view/gs/tekton-prow/pr-logs/pull/tektoncd_pipeline/5647/pull-tekton-pipeline-unit-tests/1583130831700889600

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [ ] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
Fix cloud event flacky unit tests by adding EventSender
```

### CODE REVIEW

**Comment by Yongxuanzhang** [comment]
cc @lbernick @vdemeester @imjasonh

**Comment by lbernick** [comment]
based on [this comment](https://github.com/tektoncd/pipeline/issues/4404#issuecomment-985656290) from @vdemeester  it seems like we're thinking of moving cloudevents functionality out of pipelines, so it might not make sense to combine these two packages into one, but I'm not sure if this comment is still true. @afrittoli can you comment on this?

File: pkg/reconciler/events/cloudevent/cloud_event_controller.go
Comment (lbernick) [inline pkg/reconciler/events/cloudevent/cloud_event_controller.go:148]: One thing that I'm noticing here is that the interface is really just a wrapper around a waitGroup, i.e. it is pretty similar to the original implementation that just used a plain waitGroup. I'm sorry if I led you down a rabbit hole with my suggestions about how this interface should be implemented, but now that I'm seeing it in a PR we might just want a wrapper around the CloudEventClient. Something like

```
type OurOwnCloudEventClient struct {
   client *cloudEvents.Client
   eventCount int
}

func initialize(ce *cloudEvents.Client) *OurOwnCloudEventClient {
   client = ce
   eventCount = 0
}

func (c *OurOwnCloudEventClient) SendCloudEventWithRetries(ctx, object) {
  <do some stuff>
  eventCount += 1
  go func() {
    c.client.Send()
  }
}

func (c *OurOwnCloudEventClient) checkEventsOrdered(t *testing.T, <some other args>) {
    eventsSent := []string{}
    for i := 0; i < c.eventCount, i ++ {
      e <- c.client.Events
      eventsSent = append(eventsSent, e)
    }
   // Compare expected list of events with actual list of events
}
```
In this example, eventCount can also be replaced with a waitgroup. WDYT?

File: pkg/reconciler/events/eventsender/events_order_check.go
Comment (Yongxuanzhang) [reply pkg/reconciler/events/eventsender/events_order_check.go:31]: I'm not sure if the fix can 100% replace the sleep here (Though I believe it is). Maybe if we could wait the fix is merged and then remove the sleep in another PR?

File: pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go
Comment (Yongxuanzhang) [reply pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:99]: oh yes!!

File: pkg/reconciler/events/cloudevent/cloud_event_controller.go
Comment (Yongxuanzhang) [reply pkg/reconciler/events/cloudevent/cloud_event_controller.go:148]: ok I will remove this

File: pkg/reconciler/events/cloudevent/cloudevent.go
Comment (Yongxuanzhang) [reply pkg/reconciler/events/cloudevent/cloudevent.go:85]: yes, thanks!

File: pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go
Comment (Yongxuanzhang) [reply pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:112]: I think we can not remove it, it is not related to the blocking, but a timer used to read from channel

File: pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go
Comment (Yongxuanzhang) [reply pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:113]: I tried this method and it probably doesn't work, in this function it will do regex matching from expected events to got events, the reason is the events from channel is much more complex so we cannot simply do diff between two list of strings.

File: pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go
Comment (lbernick) [reply pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:112]: Why do the tests hang if the timer is not used? The only reason this should happen is if the expected number of events is less than the actual number of events received.

File: pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go
Comment (lbernick) [reply pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:113]: Can you try defining your own comparison function? Here's an [example](https://github.com/tektoncd/pipeline/blob/main/pkg/internal/computeresources/compare/compare.go#L56-L79)

File: pkg/reconciler/events/cloudevent/cloudevent.go
Comment (lbernick) [inline pkg/reconciler/events/cloudevent/cloudevent.go:89]:
```suggestion
	// addCount increments the count of events to be sent
	addCount()
	// decreaseCount decrements the count of events to be sent, indicating the event has been sent
	decreaseCount()
}
```

File: pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go
Comment (lbernick) [inline pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:91]: I wonder if `decreaseCount` is even necessary? you could call c.waitGroup.Done() within`Send`

Comment (Yongxuanzhang) [reply pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:91]: Probably not? Because the fakeclient `Send` is not only used in `SendCloudEventWithRetries`'s gorountine. There is another function `SendCloudEvents` also used it and it doesn't need the waitgroup to block.

File: pkg/reconciler/events/cloudevent/cloudevent.go
Comment (Yongxuanzhang) [reply pkg/reconciler/events/cloudevent/cloudevent.go:89]: Thanks!

File: pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go
Comment (Yongxuanzhang) [reply pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:113]: It basically means we will reuse the current code to compare the two list? If so I don't think it adds any values here. And it may has less readability than the current version.

File: pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go
Comment (Yongxuanzhang) [reply pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:112]: It's because the previous code will loop +1 times, and during the extra loop since we don't have any events from channel, and also we don't have a timer, it will just hand there,
```
loop len(expected) + 1 times to catch extra erroneous events received that the test is not expecting
```

Comment (Yongxuanzhang) [reply pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:112]: To fix it we can just reduce the loop to len(expected) not len(expected)+1. But it would be a change to this function's behaviour. I don't know why or who implement this logic. I suggest we if we want to do it, we should leave it to another clean up PR to avoid extra changes in this PR

Comment (lbernick) [reply pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:112]: Oh I see-- the reason is because we don't want to make sure extra events have been sent, so we wait a little extra to see if any more events are received. What we really need is for the fake client to be able to indicate that it's done sending events (e.g. by closing the channel), but the fake client doesn't know when events are done being sent because they can only be sent one at a time so I'm not sure if this is possible. Let me think about if there's a better way to do this. Keeping this logic here means that our tests are still not deterministic.

Comment (Yongxuanzhang) [reply pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:112]: Sure! Thanks!

Comment (lbernick) [reply pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:112]: ok! I have an idea! We should update [newFakeClient](https://github.com/tektoncd/pipeline/blob/14eba83bf31277dc3f8c91fa798b38e30d0a8e6a/pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go#L43-L48) to initialize the fake client buffer size with len(wantEvents). 

The reason is that sends block on a buffered channel when it is full. We are currently waiting (via waitGroup.Wait()) until all events are sent before reading any events from the channel. This means if the controller sends more events than expected, these sends will block because the channel is full, and Wait() will wait indefinitely.

With this solution, if the client sends more or fewer events than expected, Wait() will block indefinitely, but if the client sends the correct number of events, Wait() will wait until they are all sent, and no longer! a.k.a. no more flakes!

Does this make sense? WDYT? If you implement this please remove the timer and add some commentary on this strategy.

Comment (Yongxuanzhang) [reply pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:112]: Maybe? Let me test this solution. Thank you for coming up with this idea!!

Comment (Yongxuanzhang) [reply pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go]: I tried to implement this in the latest commit, plz lmk if that is aligned what you're suggesting.

**Review by lbernick — APPROVED** [review]
Thanks @Yongxuanzhang for addressing this flake that has been annoying everyone since... before I joined Tekton :) Approving since this fixes the flake, and the codebase is improved by this change, but please make sure to address the readability and style comments before merging.

File: pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go
Comment (Yongxuanzhang) [reply pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:112]: Sure! Let me give it a try. Thanks for your advice!

File: pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go
Comment (Yongxuanzhang) [reply pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:91]: I'm not sure since that's confusing. 😂 We don't have gorontine in `SendCloudEvents `

File: pkg/reconciler/events/cloudevent/cloudevent.go
Comment (Yongxuanzhang) [reply pkg/reconciler/events/cloudevent/cloudevent.go:92]: Sure!

File: pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go
Comment (Yongxuanzhang) [reply pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:48]: yesm thanks!

File: pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go
Comment (Yongxuanzhang) [reply pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:93]: sure!

File: pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go
Comment (Yongxuanzhang) [reply pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:116]: removed! Thanks!

File: pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go
Comment (Yongxuanzhang) [reply pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:115]: I think the possible issue is that, we cannot guarantee that the two sorted lists have the same order. 
e.g.:
[a,b,c,] -> [a,b,c]
[za,by,xc]->[by,xc,za]
If we iterate them and compare on by one then this is not working.

File: pkg/reconciler/events/cloudevent/cloudeventclient.go
Comment (afrittoli) [inline pkg/reconciler/events/cloudevent/cloudeventclient.go:71]: NIT: CloudClient is a wrapper of CloudEvents client and implements addCount and decreaseCount

File: pkg/reconciler/events/cloudevent/cloudevent.go
Comment (afrittoli) [inline pkg/reconciler/events/cloudevent/cloudevent.go:92]: NIT: perhaps it's worth commenting that the extra methods are for testing purposes only

File: pkg/reconciler/customrun/customrun_test.go
Comment (afrittoli) [inline pkg/reconciler/customrun/customrun_test.go:194]: There is no `err` returned anymore to check!

File: pkg/reconciler/customrun/customrun_test.go
Comment (afrittoli) [inline pkg/reconciler/customrun/customrun_test.go:203]: There is no `err` returned anymore to check!

File: pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go
Comment (afrittoli) [inline pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:49]: What happens if the controller tried to send more events than expected? Would the test fail?

File: pkg/reconciler/events/cloudevent/cloudeventclient.go
Comment (Yongxuanzhang) [reply pkg/reconciler/events/cloudevent/cloudeventclient.go:71]: Thank you!

File: pkg/reconciler/events/cloudevent/cloudevent.go
Comment (Yongxuanzhang) [reply pkg/reconciler/events/cloudevent/cloudevent.go:92]: Sure!

File: pkg/reconciler/customrun/customrun_test.go
Comment (Yongxuanzhang) [reply pkg/reconciler/customrun/customrun_test.go:194]: thanks! I have fixed all of them

File: pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go
Comment (Yongxuanzhang) [reply pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:49]: Yes it will fail and this is also the previous behaviour.

Comment (Yongxuanzhang) [reply pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:49]: Oh I see the problem, this will lead to deadlock, I added a buffer size check before sending the event, if extra event wants to be sent it will return error. Plz lmk if that works, thanks!

Comment (afrittoli) [reply pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:49]: OK - it sounds good - it's important for the test to fail in that case otherwise we would miss error conditions.
Since the test method there is not trivial, I wonder if we should have a test that verifies this case.

Comment (Yongxuanzhang) [reply pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:49]: Sure! I will add a test for this!

Comment (Yongxuanzhang) [reply pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:49]: Added a test for this, ptal! 🙏

**Comment by lbernick** [comment]
/hold cancel
since this is ready for review

**Comment by lbernick** [comment]
@Yongxuanzhang could you please update your commit message title + body to match the PR description?

**Comment by Yongxuanzhang** [comment]
> @Yongxuanzhang could you please update your commit message title + body to match the PR description?

Oh thanks! I just updated it. Sorry I made some changes after @afrittoli's reviews. you may want to remove your approval and review again? 

1. I added a check in `Send` so if the channel is full we will return error, this should prevent the extra events.
2. I loop the channel size not the len(expectedEvents), this will avoid blocking on reading more events when we send less events. This should work in this PR since we will wait to ensure no events are missed

File: pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go
Comment (lbernick) [inline pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:118]: It might make sense to merge this function with `CheckCloudEventsUnordered`, since it's not used anywhere else as far as I can tell, and it needs the call to waitgroup.Wait in order to work correctly.

File: pkg/reconciler/events/cloudevent/cloudeventsfakeclient_test.go
Comment (lbernick) [inline pkg/reconciler/events/cloudevent/cloudeventsfakeclient_test.go:57]: nit: this doesn't need to be a slice

File: pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go
Comment (lbernick) [inline pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:121]: This comment is slightly misleading. Buffered channels can send once events have been pulled off of them. The reason no extra events will be sent is because the fake client returns an error in this case, but that logic lives far away from here. 

I'd suggest adding commentary to the fake client where you return an error (to explain why this prevents extra events from being sent) and some commentary here about why this block detects the case where too few events are sent.

File: pkg/reconciler/events/cloudevent/cloudeventsfakeclient_test.go
Comment (Yongxuanzhang) [reply pkg/reconciler/events/cloudevent/cloudeventsfakeclient_test.go:57]: 👍

File: pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go
Comment (Yongxuanzhang) [reply pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:118]: 👍

File: pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go
Comment (Yongxuanzhang) [reply pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:121]: 👍

File: pkg/reconciler/events/cloudevent/cloudeventsfakeclient_test.go
Comment (afrittoli) [inline pkg/reconciler/events/cloudevent/cloudeventsfakeclient_test.go:72]: Thank you 🙏

File: pkg/reconciler/testing/logger.go
Comment (afrittoli) [inline pkg/reconciler/testing/logger.go:31]: NIT: Why does this need to be a separate function?

File: test/events/events.go
Comment (afrittoli) [inline test/events/events.go:35]: NIT: It's a bit inconsistent now that the unordered check has been moved inside to the fake client but the ordered one is in this test module. It might we worth keeping this logic in the same place, for maintenance and to help test developers. BTW, do you think the ordered check suffers from the same flake as the other one?

Comment (afrittoli) [reply test/events/events.go:35]: This is something we could fix as a follow-up.

**Review by afrittoli — COMMENTED** [review]
Thank you!
/lgtm

File: test/events/events.go
Comment (Yongxuanzhang) [reply test/events/events.go:35]: Yes I'm looking into the k8s event as well, but I don't think they are the same root cause. 
Some prs:
https://github.com/tektoncd/pipeline/pull/5752
https://github.com/tektoncd/pipeline/pull/5770

**Pull request merged**
